### ISSUE TITLE
Recommendations without score

### ISSUE THREAD

**Issue by philippeluickx** [fcurella/django-recommends#24]
Hey!

I am trying to get recommendations based on behaviour, without the score aspect to it. Imagine "what other people bought/liked/read/...". In my case, I use Bookmarks.

In order to do this, I tried a few things in my def get_rating_score(self, rating):
- return 1: makes all scores in Similarities 1 as well. Which is not what I want, because some similarities are stronger, based on the amount of bookmarks that include p1 and p2
- return count of object of the current Bookmark: I got some interesting results which are already closer to what I expected. However, this drags down a score just based on the amount of Bookmarks a certain User has. Which is not what I expect... So when I have a perfect 1.0 score between 2 objects, it will plunge if one new user create a Bookmark for just one of those objects...

Maybe I am totally approaching this the wrong way, any help is appreciated!

**Comment by fcurella** [comment]
hi @philippeluickx 

I'm not sure I have enough context to answer your question.

> some similarities are stronger, based on the amount of bookmarks that include p1 and p2

what are p1 and p2 in this case?

> return count of object of the current Bookmark: [...] However, this drags down a score just based on the amount of Bookmarks a certain User has. 

Maybe you could normalize the count? ie something like `bookmark.count / user.bookmarks.count()` ?

**Comment by philippeluickx** [comment]
Hi @fcurella !
Sorry for being vague... Here's some more info:

User can browse through Companies and Bookmark those they like.
If user1 has bookmarked companyA and companyB, and user2 has bookmarked companyA (or is looking at it), then user2 should get the recommendation of companyB.

There is no rating involved, just purely taking the action of creating a few Bookmarks should be enough for giving a recommendation...

The whole recommendation thing is still pretty new for me and it gets complex very fast, so I am not sure what exactly I should be expecting. In the end, it's also not necessarily the actual score that matters, but the ranking (as long it has the HIGHEST score of the scope, it's good).

Normalizing is what I am playing around with, but can't really make any business-sense out of it.
There are 3 parameters we can play with
1. total amount of bookmarks
2. amount of bookmarks from company
3. amount of bookmarks from user

**Comment by fcurella** [comment]
> return 1: makes all scores in Similarities 1 as well. Which is not what I want, because some similarities are stronger, based on the amount of bookmarks that include p1 and p2

I _think_ you should `return 1`, but I'm still not clear on what you mean by 'some similiarities are stronger'. Can you give me some example of how they can be stronger? Is it something quantifiable? If it is, you should factor that in `get_rating_score()`

**Comment by philippeluickx** [comment]
Scenario 1:
user1 - companyA
user1 - companyB

user2 - companyA
user2 - companyB

>> similarity companyA - companyB = 1

Scenario 2:
user1 - companyA
user1 - companyB
user1 - companyC

user2 - companyA
user2 - companyB
user2 - companyC

user3 - companyA
user3 - companyB

>> similarity companyA - companyB > similarity (companyA or companyB) - companyC
maybe?
similarity companyA - companyB = 0.66 (or still 1?)
similarity companyA - companyC = 0.33
similarity companyB - companyC = 0.33

So in my case, there should be a clear connection between the AMOUNT of bookmarks and the similarity score.

Might be I do not need a recommendation engine. I could imagine a simpler implementation like:
usersA = all users that bookmarked companyA
usersA.bookmarks > group_by company and then highest count
But I figured this implementation will then only suggest the biggest companies with the highest amount of bookmarks. So not normalized...

**Comment by philippeluickx** [comment]
Well... If I use the count of bookmarks for each company, it gets me following in Scenario 2:
	companyA	1	companyB	1	1.0
	companyB	1	companyA	1	1.0
	companyC	1	companyA	1	0.414213562373
	companyC	1	companyB	1	0.414213562373
	companyA	1	companyC	1	0.414213562373
	companyB	1	companyC	1	0.414213562373

This scenario kind of makes sense, let me try some more things...

**Comment by fcurella** [comment]
I think you may be confusing the _rating score_ with the _similarity index_.

The rating score is how much a single 'vote' should count. think of it in the terms of stars for movie reviews. You need to account for the fact that a 5 store is stronger than 1 star review.

The numbers you are talking about are the similarity index. This index indicates how much an item (a company, in your case) relates to another item.

The similarity index is calculated by the _algorithm_ (http://django-recommends.readthedocs.org/en/latest/algorithms.html). There are many different way of calculating this index (in fact, there's an entire field of study and research dedicated to it), which is why I made them swappable.


To summarize, I still think you should always return 1 in `get_rating_score`.

The algorithm that you choose (by default, http://django-recommends.readthedocs.org/en/latest/algorithms.html#ghettoalgorithm) should already take care of computing the correct _similarity index_. You can see the specific lines here https://github.com/fcurella/django-recommends/blob/develop/recommends/algorithms/ghetto.py#L55

If you're not getting the expected results, you may want to replace it with a different one. You could just create your own subclass and override the methods you need.

**Comment by philippeluickx** [comment]
Fair point, this makes sense.
But damn, algorithm is a scary word...

I presume that this is the part of the code I will have to play with?

https://github.com/fcurella/django-recommends/blob/develop/recommends/algorithms/ghetto.py#L54

                        scores[item2] += similarity * rating

                        # Sum of all the similarities
                        totalSim[item2] += similarity

            # Divide each total score by total weighting to get an average
            rankings = ((item, (score / totalSim[item])) for item, score in scores.items() if totalSim[item] != 0)

**Comment by fcurella** [comment]
correct. Another bit you may want to play with is the _similarity distance_: https://github.com/fcurella/django-recommends/blob/develop/recommends/algorithms/ghetto.py#L11

You may want to try switching to the Pearson coefficient: https://github.com/fcurella/django-recommends/blob/develop/recommends/similarities.py

Note that the default algorithm is pretty slow and it's shipped only as a convenience to help you getting started. Most likely at some point in the future you will want to switch to something more refined, both mathematically and performance-wise.

**Comment by philippeluickx** [comment]
Is it ok if I keep this ticket open and share whatever I come up with later?

It might be interesting to provide also some more documentation about algorithms. What would be a scalable implementation, what are some common use cases,...?

**Comment by philippeluickx** [comment]
Ok, so it actually does seems that recsys http://django-recommends.readthedocs.org/en/latest/algorithms.html#recsysalgorithm is pretty much in line with what I was expecting. Had some issues getting it installed though...

I'll close the ticket now. Would it btw make sense to
1. provide some info on getting recsys installed
> sudo apt-get install libblas-dev liblapack-dev libatlas-base-dev gfortran
> pip install scipy
> pip install git+https://github.com/ocelma/python-recsys.git
2. provide some scenarios and results for both algorithms. Explain the differences, so people get a bit of an idea how both work before diving into it. Maybe link to some of the theory about this whole thing?
